### ISSUE TITLE
Fix single-click/double-click behaviour of folders

### ISSUE THREAD

**Issue by aral** [elementary/files#1800]
## What happens

Select/open gestures are inconsistent between folders and files (single click to open a folder, double-click to open a file).

This has different implications for different views.

### In grid view:

  - No way to select a folder using the mouse without using a workaround as clicking it will open it.

Workaround:
 
  - Click to enter folder
  - Hit back
  - The folder is now selected

Workaround 2:
 
  - Make a selection with the mouse even if you want to select just one folder.

Workaround 3 (hidden):

  - Use arrow keys to initialise a selection and use the keyboard from there.

None of these workarounds is a substitute for being able to select a folder with your mouse using a single click.

## List Mode

Single clicking a folder has the same effect as in grid mode.

It’s even more jarring here because you could open the folder in place as it is a tree. Instead, the context is lost again and the same workarounds have to be used to select a folder. 

Just as with Grid mode, this also breaks the ability to multiple select using the mouse and modifier keys.

## Columns Mode

Inconsistently with List Mode, in Columns Mode, when you select a folder, the context is not lost and it is actually possible to make a multiple selection using the mouse and modifier keys.

This is the only view where the current controls are usable.

## Expected Behavior

Folder selection via single click should be possible in every view. The controls across views should be consistent in their behaviour.

## Proposal

I feel the issue here is that gesture consistency has been favoured over effect consistency.

What I mean by that is that the Grid view and the List and Columns views are very different in terms of their handling of context. In the Grid view, “entering” a folder removes the current context within the file hierarchy. You must use the Back button to get back to the previous state. In the List and Columns views, this is not the case.

So I propose the following:

### In grid view:

  - Single click selects a folder, it does not open it.
  - Double-click opens it.
  - File selection and open/launch gestures stay the same.

Opening a folder in Grid view causes a contextual change. It makes sense for this be a non-trivial gesture. Also, this is the only way to implement mouse selection of a folder.

### In List view:

  - Single click reveals the contents of the folder in the tree layout without losing the context.
  - Double-clicking on a folder should have no effect (it should be interpreted as two single clicks). Displaying just that portion of the tree on double-click is an unexpected behaviour. If I’m using the tree view, I want to keep the hierarchy. Otherwise I’d use the Grid view.
  - File selection and open/launch gestures stay the same.

(

### In Columns view:

  - Like in List view, double-clicking on a folder should have no effect (it should be interpreted as two single clicks). Removing the context of the columns and displaying just that column on double-click is an unexpected behaviour that removes the primary utility of the Column view, which is to keep the topological relationship of the hierarchy visible.
  - Otherwise, all gestures stay the same.

### Logs

n/a

## Platform Information

```
aral@LabTop-4d073706 
-------------------- 
OS: elementary OS 6 Odin x86_64 
Host: LabTop Mk IV 
Kernel: 5.11.0-25-generic 
Uptime: 13 hours, 34 mins 
Packages: 2000 (dpkg), 28 (flatpak) 
Shell: fish 3.3.1 
Resolution: 1920x1080, 3840x2160 
DE: Pantheon 
WM: Mutter(Gala) 
Terminal: io.elementary.t 
CPU: Intel i7-10710U (12) @ 4.700GHz 
GPU: Intel Device 9bca 
Memory: 3670MiB / 15875MiB 
```

**Comment by abodedis** [comment]
I'm huge "grid view" user on elementary OS and I completely agree with you that there should be an option to double click on folders to open them and single click to select.  

"Single click to open folders" is interesting and arguably gets things done faster once you get used to it, however, I am still working on getting used to it, and am afraid I never will.

One workaround I have used for now to select a folder in the grid view using a single click: simply click the "check" on the upper left of the file.  You may have discovered this by now.

Also for List view, you also may have noticed by now the "triangle" on the left of each folder, which will reveal the contents of the folder in a "tree" view when single clicked.

Hope that helps a bit.  I'm also waiting on more options or tweaks available to change some of this behavior.

**Comment by abodedis** [comment]
I'm a huge "grid view" user on elementary OS and I completely agree with you that there should be an option to double click on folders to open them and single click to select.  

"Single click to open folders" is interesting and arguably gets things done faster once you get used to it, however, I am still working on getting used to it, and am afraid I never will.

One workaround I have used for now to select a folder in the grid view using a single click: simply click the "check" on the upper left of the file.  You may have discovered this by now.

Also for List view, you also may have noticed by now the "triangle" on the left of each folder, which will reveal the contents of the folder in a "tree" view when single clicked.

Hope that helps a bit.  I'm also waiting on more options or tweaks available to change some of this behavior.

**Comment by jeremypw** [comment]
This "mixed-mode" behaviour is a deliberate design decision the reasons for which have been documented elsewhere -  and, like anything new, it takes some getting used to.

> No way to select a folder using the mouse without using a workaround as clicking it will open it

The expected way to select a folder with the mouse in grid view is either click on the selection helper or drag a rubberband around it.  You can also click on it while holding down the `<Ctrl>` key.  None of the workrounds you mention are required.

You can select a folder in ListView by clicking anywhere except the filename column or by rubberbanding.

ColumnView is mainly intended for drilling down into a folder hierarchy so selecting a folder is a little harder - you have to use the helper or hold down the `<Ctrl` key

**Comment by aral** [comment]
Hi @jeremypw,

I’ve outlined the shortcomings of the current design in my post, above, giving you first-principles reasons for how they fall short on basic usability factors as well as how those shortcomings can be addressed by evolving the design decisions.

The issues I’ve raised are problems with the current design. Learning to work around the problems isn’t really a solution.

That said, please feel free to close this ticket if this is a “won’t fix.”

**Comment by abodedis** [comment]
I am sure that this was (and perhaps always will be) controversial for elementary OS.

It must be said, though, that it may be difficult to market elementary OS as a ["**replacement** for Windows and MacOS"](https://elementary.io/) if one of the first discussion points is that there's no option to double click to open folders.  This is basic behavior for those operating systems, and it is what a vast number of conventional, non-touch screen computer users have permanently ingrained into muscle memory.  Simply saying "get used to it, it's our design philosophy" may just drive users away, given the number of other distros available.

Having said that, I do really like using elementary OS overall.  It's my daily driver, and I'm a power user.  Hopefully other people will discover how great it is and not let these kinds of things dissuade them.

**Comment by ibrahimrahman** [comment]
The mixed-mode experience has not been nice to say the least. One argument given for it is that everything else in the OS follows a single click convention, so files should as well. What gets overlooked here is the fact that in files, a lot of operations are done on folders (selection, cut/copy/paste, renaming, multi-selection etc.) and these benefit greatly from a double click mode. I documented different issues that arise due to mixed-mode here, as well as how a double click mode resolves them, here: 

https://github.com/elementary/files/discussions/1729

I will admit, I did not have the motivation to break down each problem into a separate issue as was suggested, but the issues still remain relevant today. I thought it might be easier to get used to once Odin came out but as of now, I still find myself wanting to go back to a double-click mode.

**Comment by jeremypw** [comment]
Perhaps "alternative to Windows and MacOS" would be more accurate if you think a "replacement" means something that works the same way in all respects (which would mean supporting two different ways of working - Windows and MacOS).  The UX team has their own ideas as to how things should work.  See https://docs.elementary.io/hig/design-philosophy.

**Comment by abodedis** [comment]
> Perhaps "alternative to Windows and MacOS" would be more accurate if you think a "replacement" means something that works the same way in all respects (which would mean supporting two different ways of working - Windows and MacOS). The UX team has their own ideas as to how things should work. See https://docs.elementary.io/hig/design-philosophy.

Now that I have revisited this conversation, "replacement" and "alternative" both sound fine.  It did essentially "replace" Windows at least for me, since I use it all the time, though it behaves nothing like Windows.  

Maybe what I wanted to say was that it could be difficult to persuade a "normal" Windows or MacOS user to try elementary, instead of the OS they are used to, when a fundamental behavior of elementary OS could already be a dealbreaker.  Not everyone has an open mind or has much time to conquer the learning curve.  They just want to install a beautiful OS and get some work done.

Also, the devs are free to use any design philosophy.  That document could be an interesting read :)

**Comment by ibrahimrahman** [comment]
>Perhaps "alternative to Windows and MacOS" would be more accurate if you think a "replacement" means something that works the same way in all respects (which would mean supporting two different ways of working - Windows and MacOS). The UX team has their own ideas as to how things should work. See https://docs.elementary.io/hig/design-philosophy.

In the end, words won't count as much as the experience itself. Even if someone doesn't actually need an alternative/replacement and just want to play around, it's the experience that will make or break it for them, not the marketing. Friends on my suggestion have installed elementary OS in the past but I cannot recommend it to them any longer, the lack of a double click being one of the main reasons for that.

Your link was a good read and you're right, the UX team's vision is what decides things. But design is still my opinion, and your opinion and the opinion of all your users. And if a lot of people are having issues with something, perhaps listening to them is not a bad idea, especially since you are creating the product for them in the first place. 

I find that single-click navigation introduces some frustrating issues and all the suggested workarounds for something so simple are counter-intuitive and cumbersome. Yes elementary is different from windows and mac and ubuntu et al but being different isn't the issue here. There will always be a learning curve. The issue is that the alternative provided takes away features without improving the experience and ends up being seen as an annoyance rather than something new

**Comment by jeremypw** [comment]
It always surprises me that single-click/double-click navigation is an issue that causes such heat, to the extent that for some it is a make or break issue.  Given that, for elementaryos, the goal is that the filemanager should never need to be used, it is not surprising perhaps that there is no appetite to support two modes and a mode was chosen that was thought to be acceptable (if not ideal) to the most users, accepting that a single mode will never please everyone.  It is perfectly acceptable to propose a better *single mode* experience but I doubt that reverting to two modes is likely at this point.

At the end of the day, there *are* distros that aim to reproduce the Windows, or MacOS, experience as exactly as possible to cater for users who are unwilling or unable to change their way of working. Elementaryos is not one of them.

**Comment by jeremypw** [comment]
There is already a discussion opened on this topic #1729 so I think I will close this issue as "Will not fix" for now. Feel free to continue the discussion in the topic mentioned.

**Comment by aral** [comment]
> Given that, for elementaryos, the goal is that the filemanager should never need to be used

This goal makes no sense unless you’re planning on creating an operating system without a hierarchical file system, where all files are essentially kept per app in a database. Now, there’s something to be said about such a system but an operating system like Linux where _everything is a file_ is not it.

Could you please let me know where this goal is articulated as I haven’t been able to find it in the Human Interface Guidelines (https://docs.elementary.io/hig/).  

> , it is not surprising perhaps that there is no appetite to support two modes and a mode was chosen that was thought to be acceptable (if not ideal) to the most users, accepting that a single mode will never please everyone. It is perfectly acceptable to propose a better _single mode_ experience but your proposal seems to be to use double-click mode which would alienate all fans single-click mode. I doubt that reverting to two modes with a hidden setting is likely at this point.

I don’t understand what modes you’re referring to here.

The fundamental flaw in this design is that it treats two very different views as the same and expects the same gestures to make sense within them.

Specifically:

Grid: a drill-down view where hierarchical context is lost when navigating between branches of the directory tree

List and Columns: a selective reveal view where hierarchical context is not lost during navigation. (Or should not be. If you double-click a folder in these views, the previous expectations are violated and it behaves as a drill-down. Since I cannot imagine this is a design decision, I’m going to assume it’s a bug.)

> At the end of the day, there _are_ distros that aim to reproduce the Windows, or MacOS, experience as exactly as possible to cater for users who are unwilling or unable to change their way of working. Elementaryos is not one of them.

Which is why I gave first principles reasons for why the interactions are broken. On the other hand, if your web site says “ethical replacement for Windows and macOS”, you might want to rethink either that copy or your approach to people who are switching from those systems. 

Design is a process. I feel there’s a danger is adhering to past decisions as dogma in the face of evidence of shortcomings and demonstrated means of improving those shortcomings.

I’ve been using graphical interfaces for over 30 years at this point. I can safely say that elementary OS Files is the first time I’ve been unable to select a folder by clicking on it. I find it rather amazing that a feature that was working correctly on the Xerox Alto in 1973 is such a contentious one in 2021.

**Comment by jeremypw** [comment]
As this issue is currently closed it would be better for visibility if you continued the discussion here: https://github.com/elementary/files/discussions/1729.

> This goal makes no sense unless you’re planning on creating an operating system without a hierarchical file system, where all files are essentially kept per app in a database.

I think the idea is that for users other than computer experts the underlying file system is of no interest and should not need to be manipulated to get work done.  Computer experts will always be able to achieve their ends, if necessary by installing another filemanager.

Discussions re design are best carried out with the UX team.  I freely admit to not being a design expert - I just fix bugs in the software where I can and try to reflect my understanding of the UX team's design philosophy.  Any changes I make that does not accord with that philosophy will not get merged.
